event/event_target_bus.go

```diff
@@ -0,0 +1,162 @@
+// licenses/APL.txt.
+
+//go:build js
+
+package event
+
+import (
+	"sync"
+	"sync/atomic"
+	"syscall/js"
+
+	"github.com/sourcenetwork/goji"
+)
+
+type eventTargetSub struct {
+	id       uint64
+	listener js.Func
+	value    chan Message
+	events   []Name
+}
+
+// Message returns the message channel for the subscription.
+func (s *eventTargetSub) Message() <-chan Message {
+	return s.value
+}
+
+// eventTargetBus uses a JavaScript EventTarget to
+// manage subscribers and publish messages.
+type eventTargetBus struct {
+	target       goji.EventTargetValue
+	bufferSize   uint
+	subscriberID atomic.Uint64
+	subscribers  map[uint64]*eventTargetSub
+	closeMutex   sync.Mutex
+}
+
+// NewEventTargetBus creates a new bus from the given event target value.
+func NewEventTargetBus(value js.Value, bufferSize uint) Bus {
+	return &eventTargetBus{
+		target:      goji.EventTargetValue(value),
+		bufferSize:  bufferSize,
+		subscribers: make(map[uint64]*eventTargetSub),
+	}
+}
+
+// Publish broadcasts the given message to the bus subscribers. Non-blocking.
+func (b *eventTargetBus) Publish(msg Message) {
+	detail := goji.MustMarshalJS(msg.Data)
+	event := goji.CustomEvent.New(string(msg.Name), detail)
+	b.target.DispatchEvent(js.Value(event))
+}
+
+// Subscribe returns a new subscription that will receive all of the events
+// contained in the given list of events.
+func (b *eventTargetBus) Subscribe(events ...Name) (Subscription, error) {
+	b.closeMutex.Lock()
+	defer b.closeMutex.Unlock()
+
+	value := make(chan Message, b.bufferSize)
+	listener := goji.EventListener(func(event goji.EventValue) {
+		value <- unmarshalMessage(event)
+	})
+
+	sub := &eventTargetSub{
+		id:       b.subscriberID.Add(1),
+		listener: listener,
+		value:    value,
+		events:   events,
+	}
+	b.subscribers[sub.id] = sub
+
+	return sub, nil
+}
+
+// Unsubscribe removes all event subscriptions and closes the subscription.
+//
+// Will do nothing if this object is already closed.
+func (b *eventTargetBus) Unsubscribe(sub Subscription) {
+	b.closeMutex.Lock()
+	defer b.closeMutex.Unlock()
+
+	s, ok := sub.(*eventTargetSub)
+	if ok {
+		b.unsubscribe(s)
+	}
+}
+
+func (b *eventTargetBus) unsubscribe(sub *eventTargetSub) {
+	if _, ok := b.subscribers[sub.id]; !ok {
+		return
+	}
+	for _, e := range sub.events {
+		b.target.RemoveEventListener(string(e), sub.listener.Value, js.Undefined())
+	}
+	close(sub.value)
+	delete(b.subscribers, sub.id)
+	sub.listener.Release()
+}
+
+// Close unsubscribes all active subscribers and closes the bus.
+func (b *eventTargetBus) Close() {
+	b.closeMutex.Lock()
+	defer b.closeMutex.Unlock()
+
+	for _, s := range b.subscribers {
+		b.unsubscribe(s)
+	}
+}
+
+// unmarshalMessage unmarshals a message from a JS EventValue.
+func unmarshalMessage(event goji.EventValue) Message {
+	message := Message{
+		Name: Name(event.Type()),
+	}
+	detail := js.Value(event).Get("detail")
+	if detail.IsUndefined() {
+		return message
+	}
+	switch message.Name {
```

Comment:
suggestion: Might be worth adding a `//no-op` `default` case here with a comment detailing why that is okay/the-lesser-evil.  And perhaps a similar comment for the `detail.IsUndefined()` early return.  Or perhaps note it in the function signature.

Without this, the reader is left guessing, and in the case of the `switch` it wouldn't be too hard to guess that it was accidental.